src/CodeGen_X86.cpp

```diff
@@ -667,6 +668,38 @@ void CodeGen_X86::visit(const Call *op) {
         }
     }
 
+    // clang-format off
+    static const Pattern reinterpret_patterns[] = {
+        {"saturating_narrow", i16_sat(wild_u32x_)},
+        {"saturating_narrow", u16_sat(wild_u32x_)},
+        {"saturating_narrow", i8_sat(wild_u16x_)},
+        {"saturating_narrow", u8_sat(wild_u16x_)},
+    };
+    // clang-format on
+
+    // Search for saturating casts where the inner value can be
+    // reinterpreted to signed, so that we can use existing
+    // saturating_narrow patterns.
+    for (const auto &pattern : reinterpret_patterns) {
+        if (expr_match(pattern.pattern, op, matches)) {
+            const Expr &expr = matches[0];
+            const Type &t = expr.type();
+            // TODO: might want to keep track of scope of bounds information.
+            const ConstantInterval ibounds = constant_integer_bounds(expr);
+            const Type reint_type = Int(t.bits());
+            if (ibounds.max_defined && reint_type.can_represent(ibounds.max)) {
```

Comment:
You already make the type t.with_code(halide_type_int) below, so I suggest making reint_type that instead (just in case vector lanes has some bearing later), and then saying reint_type.can_represent(ibounds) instead so that the reader doesn't have to think about whether the min needs checking too.

Reply:
`reint_type.can_represent(ibounds)` requires that the interval is bounded, but I only need an upper bound here (and only need a lower bound below), so I think that is less powerful than here. I will fix the `t.with_code(halide_type_int)` stuff though.

Reply:
constant_integer_bounds uses the bounds of the type if it can't find something better, so constant_integer_bounds(some_int_or_uint_expr) always returns a defined upper and lower bound, with one exception: If it's a uint64, the upper bound may not be defined.

It's only really when you start doing math on the ConstantInterval objects themselves (which represent infinite integers) that you get undefined upper or lower bounds.